File: solutions/scala/051-set-matrix-zeroes.scala
/**
 * Problem 51: Set Matrix Zeroes (LeetCode 73)
 * Difficulty: Med
 * Language: Scala
 */
def setZeroes(matrix: Array[Array[Int]]): Array[Array[Int]] = {
  val zeroRows = matrix.indices.filter(i => matrix(i).contains(0)).toSet
  val zeroCols = matrix(0).indices.filter(j => matrix.exists(_(j) == 0)).toSet
  matrix.zipWithIndex.map { case (row, i) =>
    row.zipWithIndex.map { case (v, j) =>
      if (zeroRows(i) || zeroCols(j)) 0 else v
    }
  }
}

File: solutions/rust/051-set-matrix-zeroes.rs
/**
 * Problem 51: Set Matrix Zeroes (LeetCode 73)
 * Difficulty: Med
 * Language: Rust
 */
pub fn set_zeroes(matrix: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
    let zero_rows: std::collections::HashSet<usize> = matrix.iter().enumerate()
        .filter(|(_, r)| r.contains(&0)).map(|(i, _)| i).collect();
    let cols = matrix[0].len();
    let zero_cols: std::collections::HashSet<usize> = (0..cols)
        .filter(|&j| matrix.iter().any(|r| r[j] == 0)).collect();
    matrix.iter().enumerate().map(|(i, row)|
        row.iter().enumerate().map(|(j, &v)|
            if zero_rows.contains(&i) || zero_cols.contains(&j) { 0 } else { v }
        ).collect()
    ).collect()
}
